docs: Adds Prometheus howto guide (#944)
canonical/k8s-snap: docs/src/snap/howto/observability.md

# How to use Prometheus with {{product}}

Observability is an essential component in any system for understanding,
managing, and improving its performance and reliability. The main pillars of
observability are metrics, logs and traces.

One of these pillars is covered by [Prometheus][Prometheus], an open-source
systems monitoring and alerting toolkit designed to collect, process, and query
time-series metrics.

This guide walks you through installing Prometheus in a {{product}} environment.

## Prerequisites

This guide assumes the following:

- You have root or sudo access to the machine.
- You have installed the {{product}} snap.
  (see How-to [Install {{product}} from a snap][snap-install-howto]).
- You have a bootstrapped {{product}} cluster (see the [Getting Started][
  getting-started-guide] guide).
- You have enabled a persistent storage solution in your cluster
  (see How-to [Enable persistent storage][enable-storage]).

## Install Prometheus

Prometheus and its operator can be installed with a Helm chart. Start by
adding the community Helm chart repository to your system:

```
sudo k8s helm repo add prometheus-community https://prometheus-community.github.io/helm-charts

> sudo k8s helm repo update

Reviewer: We don't want to reference/use the k8s helm command because it creates dependencies that we might not want to have in the future. 
Please link to the helm tutorial https://github.com/canonical/k8s-snap/pull/964/files#diff-f78900c9074145b3ce76e55215bcbfeb7c34268400df46895750d6ed6f4041f8R9 or to upstream on how to install helm

file. You can generate it by running:

```
sudo k8s helm show values prometheus-community/kube-prometheus-stack > values.yaml
```

In order to ensure High Availability for the Prometheus services, make sure to
configure the `volumeClaimTemplate` sections appropriately for Alertmanager and
Prometheus (and ThanosRuler, if enabled), and the `persistence` section for
Grafana. If the `storageClassName` field is not set, the cluster's default
`StorageClass` will be used instead. Make sure you have a `StorageClass`
created for your persistent storage solution of choice. You can list them by
running the following:

```
sudo k8s kubectl get storageclass
```

After the Prometheus deployment has been customized with the
`values.yaml` file, run the following command:

```
sudo k8s helm install prometheus prometheus-community/kube-prometheus-stack \
  --create-namespace --namespace observability -f values.yaml
```

Note that this Helm chart installs a few dependent charts:

- `prometheus-community/kube-state-metrics`
- `prometheus-community/prometheus-node-exporter`
- `grafana/grafana`

## Verify that Prometheus is running

> It is recommended to ensure that Prometheus initialises properly and is running

Reviewer: nit: initializes (we moved to US English recently)

logs, and traces.

If you've deployed Prometheus with the Helm chart above, you should already
have Grafana deployed in your cluster:

```
sudo k8s kubectl get pods -n observability -l "app.kubernetes.io/name=grafana"
```

Next, connect to the Grafana dashboard through its Kubernetes service:

```
SVC_NAME="prometheus-grafana"
sudo k8s kubectl get -n observability svc/$SVC_NAME
CLUSTER_IP="$(sudo k8s kubectl get -n observability svc/$SVC_NAME -o jsonpath='{.spec.clusterIP}')"
CLUSTER_IP_PORT="$(sudo k8s kubectl get -n observability svc/$SVC_NAME -o jsonpath='{.spec.ports[0].port}')"
echo "Grafana dashboard URL (ClusterIP): http://${CLUSTER_IP}:${CLUSTER_IP_PORT}/"
```

If you do not have access to the cluster network, or if the Grafana Kubernetes
service is not exposed externally, you can instead create a temporary local
port-forward to the Grafana dashboard:

```
export POD_NAME=$(sudo k8s kubectl get pods --namespace observability -l "app.kubernetes.io/name=grafana" -o jsonpath="{.items[0].metadata.name}")
sudo k8s kubectl --namespace observability port-forward $POD_NAME 3000
```

The default username/password for Grafana are: `admin`/`prom-operator`

> # Removing Prometheus

Author: Done

Prometheus and its related components (including Grafana) can be removed by
running:

```
sudo k8s helm delete prometheus -n observability
```

> **_NOTE:_**: The Persistent Volumes created for Prometheus and its related
> services may not deleted when removing Prometheus. You can check them
> by running:

```
sudo k8s kubectl get -n observability pvc
sudo k8s kubectl get pv
```

<!-- LINKS -->

[Prometheus]: https://prometheus.io/
[snap-install-howto]: ./install/snap.md

> [getting-started-guide]: ../../tutorial/getting-started.md

Author: Done

# How to use Prometheus with Canonical K8s

Observability is an essential component in any system for understanding,
managing, and improving its performance and reliability. The main pillars of

> observability are: metrics, logs, traces.

Reviewer 2: ```suggestion
observability are metrics, logs and traces.
```

One of these pillars is covered by [Prometheus][Prometheus], an open-source
systems monitoring and alerting toolkit designed to collect, process, and query
time-series metrics.

This guide walks you through installing Prometheus in a {{product}} environment.

> ## What you'll need

Reviewer 2: ```suggestion
## Prerequisites
```

This guide assumes the following:

- You have root or sudo access to the machine.
- You have installed the {{product}} snap.
  (see How-to [Install {{product}} from a snap][snap-install-howto]).
- You have a bootstrapped {{product}} cluster (see the [Getting Started][
  getting-started-guide] guide).
- You have enabled a persistent storage solution in your cluster
  (see How-to [Enable persistent storage][enable-storage]).

## Install Prometheus

> Prometheus and its operator can be installed through a Helm chart. Start by

Reviewer 2: ```suggestion
Prometheus and its operator can be installed with a Helm chart. Start by
```

adding the community Helm chart repository to your system:

```bash
sudo k8s helm repo add prometheus-community https://prometheus-community.github.io/helm-charts
sudo k8s helm repo update
```

> Before deploying the Helm chart, you can customize it through a `values.yaml`

Reviewer 2: ```suggestion
Before deploying the Helm chart, you can customize it with a `values.yaml`
```

file. You can generate it by running:

```bash
sudo k8s helm show values prometheus-community/kube-prometheus-stack > values.yaml
```

In order to ensure High Availability for the Prometheus services, make sure to
configure the `volumeClaimTemplate` sections appropriately for Alertmanager and
Prometheus (and ThanosRuler, if enabled), and the `persistence` section for
Grafana. If the `storageClassName` field is not set, the cluster's default
`StorageClass` will be used instead. Make sure you have a `StorageClass`
created for your persistent storage solution of choice. You can list them by

> running the following:

Reviewer 2: I'm not familiar with volumeClaimTemplate, but I don't feel like this paragraph clearly explains how this is related to HA.

Author: Right. By default and from what I've seen, this Helm chart won't create and PVCs / PVs for Prometheus, Grafana, Alertmanager. By configuring the `volumeClaimTemplate`, it will then create volumes for it. This is very important, as without this, these applications could lose their data if the Pod gets destroyed / recreated as their storage is ephemeral.

As for the high availability part, and this could be included in the explanation maybe, if for some reason a node crashes and doesn't recover, the application also goes down. If you don't have PV / PVCs, Kubernetes may recreate those Pods on another node, and they won't have any of the previously collected data. If you're using persistent volumes from (and ofcourse, non-local PVs like `rawfile-localpv` - if it's a local volume, the Pods might not be recreated on another node as the PVs are going to have a node affinity set to the original node, preventing high availability), this issue is avoided, allowing the Pods to be recreated on other nodes with their storage intact.

```bash
sudo k8s kubectl get storageclass
```

> After the Prometheus deployment has been customized appropriately through the

Reviewer 2: ```suggestion
After the Prometheus deployment has been customized appropriately with the
```

Author: Done

`values.yaml` file, run the following command:

```bash
sudo k8s helm install prometheus prometheus-community/kube-prometheus-stack \
  --create-namespace --namespace observability -f values.yaml
```

Note that this Helm chart installs a few dependent charts:

- `prometheus-community/kube-state-metrics`
- `prometheus-community/prometheus-node-exporter`
- `grafana/grafana`

Visit https://github.com/prometheus-operator/kube-prometheus for instructions
on how to configure the Alertmanager and Prometheus instances using the

> Prometheus Operator.

Reviewer 2: Can you link to the exact page?

Author: I'll remove this, as this is also linked by the helm chart after installing anyways.

## Verify that Prometheus is running

It is recommended to ensure that Prometheus initialises properly and is running

> without issues. Check that the Prometheus Pods are running:

Reviewer 2: ```suggestion
without issues. Check that the Prometheus pods are running:
```

Author: Done

```bash
sudo k8s kubectl get pods -n observability -l "app.kubernetes.io/name=prometheus"
```

Next, connect to the Prometheus dashboard through its Kubernetes Service:

```bash
SVC_NAME="prometheus-kube-prometheus-prometheus"
sudo k8s kubectl get -n observability svc/$SVC_NAME
CLUSTER_IP="$(sudo k8s kubectl get -n observability svc/$SVC_NAME -o jsonpath='{.spec.clusterIP}')"
CLUSTER_IP_PORT="$(sudo k8s kubectl get -n observability svc/$SVC_NAME -o jsonpath='{.spec.ports[0].port}')"
echo "Prometheus dashboard URL (ClusterIP): http://${CLUSTER_IP}:${CLUSTER_IP_PORT}/graph"
```

> If you do not have access to the cluster network, and if the Prometheus

Reviewer 2: ```suggestion
If you do not have access to the cluster network, or if the Prometheus
```

Author: Done

Kubernetes service is not exposed externally, you can instead create a
temporary local port-forward to the Prometheus dashboard:

```bash
export POD_NAME=$(sudo k8s kubectl get pods --namespace observability -l "app.kubernetes.io/name=prometheus" -o jsonpath="{.items[0].metadata.name}")
sudo k8s kubectl --namespace observability port-forward $POD_NAME 9090
```

> You can check the metrics scraped by Prometheus by running:

Reviewer 2: ```suggestion
You can check the metrics scraped that have been scraped so far by running:
```

Author: Done

```bash
curl -s http://${CLUSTER_IP}:${CLUSTER_IP_PORT}/metrics
```

## Accessing Grafana

[Grafana][Grafana] is an open-source analytics and visualization web
application that enables you query, visualize, alert on, and explore metrics,
logs, and traces.

If you've deployed Prometheus with the Helm chart above, you should already

> have deployed in your cluster:

Author: Done

```bash
sudo k8s kubectl get pods -n observability -l "app.kubernetes.io/name=grafana"
```

> Next, connect to the Grafana dashboard through its Kubernetes Service:

Reviewer 2: ```suggestion
Next, connect to the Grafana dashboard through its Kubernetes service:
```

Author: Done

```bash
SVC_NAME="prometheus-grafana"
sudo k8s kubectl get -n observability svc/$SVC_NAME
CLUSTER_IP="$(sudo k8s kubectl get -n observability svc/$SVC_NAME -o jsonpath='{.spec.clusterIP}')"
CLUSTER_IP_PORT="$(sudo k8s kubectl get -n observability svc/$SVC_NAME -o jsonpath='{.spec.ports[0].port}')"
echo "Grafana dashboard URL (ClusterIP): http://${CLUSTER_IP}:${CLUSTER_IP_PORT}/"
```

> If you do not have access to the cluster network, and if the Grafana Kubernetes

Reviewer 2: ```suggestion
If you do not have access to the cluster network, or if the Grafana Kubernetes
```

Author: Done

service is not exposed externally, you can instead create a temporary local
port-forward to the Grafana dashboard:

```bash
export POD_NAME=$(sudo k8s kubectl get pods --namespace observability -l "app.kubernetes.io/name=grafana" -o jsonpath="{.items[0].metadata.name}")
sudo k8s kubectl --namespace observability port-forward $POD_NAME 3000
```

> The default username/password for Grafana are: `admin/prom-operator`

Reviewer 2: ```suggestion
The default username/password for Grafana are: `admin`/`prom-operator`
```

Author: Done

# Removing Prometheus

> Prometheus and its related components can be removed by running:

Author: It is also going to be removed automatically with Prometheus.

```bash
sudo k8s helm delete prometheus -n observability
```

> **_NOTE:_**: The Persistent Volumes created for Prometheus and its related

> > services may not deleted when when removing Prometheus. You can check them

Reviewer 2: ```suggestion
> services may not deleted when removing Prometheus. You can check them
```

Author: Done
